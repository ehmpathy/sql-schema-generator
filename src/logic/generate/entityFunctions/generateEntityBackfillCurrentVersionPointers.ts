import indentString from 'indent-string';

import { Entity } from '../../../domain';

/*
1. define procedure / function (try function if possible)
2. logic
  1. insert the entities that don't have a CVP
  2. update the entities that do have a CVP but are out of sync

note: insert on conflict update was not used because there would be no efficient way to determine number of rows affected
  - since "on conflict update" shows as "two" rows affected... https://stackoverflow.com/questions/3747314/why-are-2-rows-affected-in-my-insert-on-duplicate-key-update
*/

export const generateEntityBackfillCurrentVersionPointers = ({ entity }: { entity: Entity }) => {
  // define select statement for getting current version for entity
  const selectCurrentVersionIdForEntity = `
SELECT
  e.id as id,
  v.id as current_version_id
FROM ${entity.name} e
JOIN ${entity.name}_version v ON v.${entity.name}_id = e.id
WHERE 1=1
  AND effective_at = ( -- and is the currently effective version
    SELECT MAX(effective_at)
    FROM ${entity.name}_version ssv
    WHERE ssv.${entity.name}_id = e.id -- for same entity
  )
  `.trim();

  // combine the version and static logic into full upsert function
  const definition = `
CREATE OR REPLACE FUNCTION backfill_${entity.name}_cvp(
  in_limit int
)
RETURNS int
LANGUAGE plpgsql
AS $$
  DECLARE
    v_cvp_rows_inserted int;
    v_cvp_rows_updated int;
    v_remaining_limit int;
  BEGIN
    -- 1. insert every time cvp dne for an entity
    INSERT INTO ${entity.name}_cvp
      (${entity.name}_id, ${entity.name}_version_id)
    SELECT
      e.id,
      e_to_cv.current_version_id
    FROM ${entity.name} e
    JOIN (
      ${indentString(selectCurrentVersionIdForEntity, 6).trim()}
    ) AS e_to_cv ON e_to_cv.id = e.id
    WHERE 1=1
      AND NOT EXISTS (
        SELECT 'x'
        FROM ${entity.name}_cvp cvp
        WHERE cvp.${entity.name}_id = e.id
      )
    LIMIT in_limit;
    GET DIAGNOSTICS v_cvp_rows_inserted = ROW_COUNT;
    v_remaining_limit := in_limit - v_cvp_rows_inserted;

    -- 2. update every time cvp exists but is out of sync for an entity
    WITH cvp_to_update AS (
      SELECT
        cvp.id,
        e_to_cv.current_version_id as actual_current_version_id
      FROM (
        ${indentString(selectCurrentVersionIdForEntity, 6).trim()}
      ) AS e_to_cv
      JOIN ${entity.name}_cvp cvp ON cvp.${entity.name}_id = e_to_cv.id
      WHERE cvp.${entity.name}_version_id <> e_to_cv.current_version_id
      LIMIT v_remaining_limit
    )
    UPDATE ${entity.name}_cvp
    SET
      updated_at = now(),
      ${entity.name}_version_id = cvp_to_update.actual_current_version_id
    FROM cvp_to_update
    WHERE ${entity.name}_cvp.id = cvp_to_update.id;
    GET DIAGNOSTICS v_cvp_rows_updated = ROW_COUNT;

    -- return the number of rows affected
    RETURN v_cvp_rows_inserted + v_cvp_rows_updated;
  END;
$$
  `.trim();
  return {
    name: `backfill_${entity.name}_cvp`,
    sql: definition,
  };
};
