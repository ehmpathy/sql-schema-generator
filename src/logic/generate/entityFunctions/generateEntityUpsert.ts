import { Entity, Property } from '../../../types';
import { extractMysqlTypeDefinitionFromProperty } from '../utils/extractMysqlTypeDefinitionFromProperty';

/*
1. define procedure / function (try function if possible)
2. define inputs
3. declare needed inputs
4. logic
  1. domain logic validation? - no, user should wrap our sproc in their own if they want custom logic - or modify the generated one
  2. find or create the static entity, get the id of the static entity
  3. check if the version has changed
  4. if the version has changed, insert a new version
  5. return the id of the static entity
*/

const propertyNameToWhereClauseConditional = ({ name, entity }: { name: string; entity: Entity }) =>
  [
    `AND (${name} = BINARY in_${name}`, // compare against binary rep of string to ensure case sensitivity (even though column may be case sensitive, mysql will still "help us out")
    entity.properties[name].nullable ? ` OR (${name} IS null AND in_${name} IS null)` : '', // NULL != NULL, so special check if field is nullable
    ')',
  ].join('');
const propertyNameToInputName = (name: string) => `in_${name}`;
const propertyToFunctionInputDefinition = ({ name, definition }: { name: string; definition: Property }) =>
  [propertyNameToInputName(name), extractMysqlTypeDefinitionFromProperty({ property: definition })].join(' ');

export const generateEntityUpsert = ({ entity }: { entity: Entity }) => {
  // define the property names
  const staticPropertyNames = Object.entries(entity.properties)
    .filter((entry) => !entry[1].updatable)
    .map((entry) => entry[0]);
  const updateablePropertyNames = Object.entries(entity.properties)
    .filter((entry) => !!entry[1].updatable)
    .map((entry) => entry[0]);

  // define where clause conditionals
  const uniqueStaticPropertyNames = staticPropertyNames.filter((name) => entity.unique.includes(name));
  const uniqueStaticPropertyWhereClauseConditionals = uniqueStaticPropertyNames.map((name) =>
    propertyNameToWhereClauseConditional({ name, entity }),
  );
  const updateablePropertyWhereClauseConditionals = updateablePropertyNames.map((name) =>
    propertyNameToWhereClauseConditional({ name, entity }),
  );

  // define the input definitions
  const inputDefinitions = Object.entries(entity.properties).map((entry) =>
    propertyToFunctionInputDefinition({ name: entry[0], definition: entry[1] }),
  );

  // define the findOrCreateStaticEntityLogic
  const findOrCreateStaticEntityLogic = `
  -- find or create the static entity
  SET v_static_id = (
    SELECT id
    FROM ${entity.name}
    WHERE 1=1
      ${uniqueStaticPropertyWhereClauseConditionals.join('\n      ')}
  );
  IF (v_static_id IS NULL) THEN -- if entity could not be found originally, create the static entity
    INSERT INTO ${entity.name}
      (uuid, ${staticPropertyNames.join(', ')})
      VALUES
      (uuid(), ${staticPropertyNames.map(propertyNameToInputName).join(', ')});
    SET v_static_id = (
      SELECT id
      FROM ${entity.name}
      WHERE 1=1
        ${uniqueStaticPropertyWhereClauseConditionals.join('\n        ')}
    );
  END IF;
  `.trim();

  // define the update version logic
  const insertVersionIfDynamicDataChangedLogic = updateablePropertyNames.length
    ? `
  -- insert new version to ensure that latest dynamic data is effective, if dynamic data has changed
  SET v_matching_version_id = ( -- see if latest version already has this data
    SELECT id
    FROM ${entity.name}_version
    WHERE 1=1
      AND ${entity.name}_id = v_static_id -- for this entity
      AND effective_at = ( -- and is the currently effective version
        SELECT MAX(effective_at)
        FROM ${entity.name}_version ssv
        WHERE ssv.${entity.name}_id = v_static_id
      )
      ${updateablePropertyWhereClauseConditionals.join('\n      ')}
  );
  IF (v_matching_version_id IS NULL) THEN -- if the latest version does not match, insert a new version
    INSERT INTO ${entity.name}_version
      (${entity.name}_id, ${updateablePropertyNames.join(', ')})
      VALUES
      (v_static_id, ${updateablePropertyNames.map(propertyNameToInputName).join(', ')});
  END IF;
    `.trim()
    : ''; // if no updateable properties, no logic needed

  const upsertCurrentVersionPointerIfNeededLogic = updateablePropertyNames.length
    ? `
  -- update the current version pointer table, if it is not already up to date
  SET v_current_version_id = ( -- get the current version id
    SELECT id
    FROM ${entity.name}_version
    WHERE 1=1
      AND ${entity.name}_id = v_static_id -- for this entity
      AND effective_at = ( -- and is the currently effective version
        SELECT MAX(effective_at)
        FROM ${entity.name}_version ssv
        WHERE ssv.${entity.name}_id = v_static_id
      )
  );
  SET v_current_version_id_recorded_in_pointer_table = ( -- get the version recorded as current for the entity, if any
    SELECT ${entity.name}_version_id
    FROM ${entity.name}_cvp
    WHERE 1=1
      AND ${entity.name}_id = v_static_id -- for this entity
  );
  IF (v_current_version_id_recorded_in_pointer_table IS null) THEN -- if its null, then just insert it, since it isn't already defined
    INSERT INTO ${entity.name}_cvp
      (${entity.name}_id, ${entity.name}_version_id)
      VALUES
      (v_static_id, v_current_version_id);
    SET v_current_version_id_recorded_in_pointer_table = v_current_version_id; -- and record that the current version recorded is now the real current version
  END IF;
  IF (v_current_version_id_recorded_in_pointer_table <> v_current_version_id) THEN -- if they are not exactly equal, update the current version recorded in the pointer tabe
    UPDATE ${entity.name}_cvp
    SET
      ${entity.name}_version_id = v_current_version_id,
      updated_at = NOW(6)
    WHERE
    ${entity.name}_id = v_static_id;
  END IF;
  `.trim()
    : undefined;

  // combine the version and static logic into full upsert function
  const definition = `
CREATE FUNCTION \`upsert_${entity.name}\`(
  ${inputDefinitions.join(',\n  ')}
) RETURNS bigint(20)
BEGIN
  -- declarations
  DECLARE v_static_id BIGINT;
  DECLARE v_matching_version_id BIGINT;
  DECLARE v_current_version_id BIGINT;
  DECLARE v_current_version_id_recorded_in_pointer_table BIGINT;

  ${[findOrCreateStaticEntityLogic, insertVersionIfDynamicDataChangedLogic, upsertCurrentVersionPointerIfNeededLogic]
    .filter((sql) => !!sql)
    .join('\n\n  ')}

  -- return the static entity id
  return v_static_id;
END
  `.trim();
  return {
    name: `upsert_${entity.name}`,
    sql: definition,
  };
};
