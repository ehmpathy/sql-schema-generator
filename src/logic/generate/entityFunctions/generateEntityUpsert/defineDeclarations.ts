import { Entity } from '../../../../types';

export const defineDeclarations = ({ entity }: { entity: Entity }) => {
  const declarations = [];

  // add the static table declaration
  declarations.push('DECLARE v_static_id BIGINT;');

  // add the dynamic version table declarations, if needed
  const hasUpdatableProperties = Object.values(entity.properties).some((property) => !!property.updatable);
  if (hasUpdatableProperties) {
    declarations.push(
      'DECLARE v_matching_version_id BIGINT;',
      'DECLARE v_current_version_id BIGINT;',
      'DECLARE v_current_version_id_recorded_in_pointer_table BIGINT;',
    );
  }

  // add the mapping table loop declarations, if needed
  const hasArrayProperties = Object.values(entity.properties).some((property) => !!property.array);
  if (hasArrayProperties) {
    declarations.push(
      'DECLARE v_delimited_string_access_index INT;',
      'DECLARE v_delimited_string_access_value VARCHAR(255);',
      'DECLARE v_can_still_find_values_in_delimited_string BOOLEAN;',
    );
  }

  // return the declarations
  return declarations;
};
