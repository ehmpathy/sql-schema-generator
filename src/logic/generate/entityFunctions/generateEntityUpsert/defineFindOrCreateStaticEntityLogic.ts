import { Entity } from '../../../../types';
import { castPropertyToColumnName } from '../../utils/castPropertyToColumnName';
import { pickKeysFromObject } from '../../utils/pickKeysFromObject';
import { defineMappingTableInsertsForArrayProperty } from './defineMappingTableInsertsForArrayProperty';
import { castPropertyToTableColumnValueReference } from './utils/castPropertyToTableColumnValueReference';
import { castPropertyToWhereClauseConditional } from './utils/castPropertyToWhereClauseConditional';

export const defineFindOrCreateStaticEntityLogic = ({ entity }: { entity: Entity }) => {
  // define the static property names
  const staticPropertyNames = Object.entries(entity.properties)
    .filter((entry) => !entry[1].updatable)
    .map((entry) => entry[0]);

  // define the column names and the column value references for the static properties
  const staticPropertyColumnNames = staticPropertyNames.map((name) =>
    castPropertyToColumnName({ name, definition: entity.properties[name] }),
  );
  const staticPropertyColumnValueReferences = staticPropertyNames.map((name) =>
    castPropertyToTableColumnValueReference({ name, definition: entity.properties[name] }),
  );

  // define the unique static property where clauses (for finding the entity by unique values)
  const uniqueStaticPropertyNames = staticPropertyNames.filter((name) => entity.unique.includes(name));
  const uniqueStaticPropertyWhereClauseConditionals = uniqueStaticPropertyNames.map((name) =>
    castPropertyToWhereClauseConditional({ name, definition: entity.properties[name] }),
  );

  // define the array properties, for which we'll need to insert into a mapping table
  const staticArrayProperties = pickKeysFromObject({
    object: entity.properties,
    keep: (property) => !!property.array && !property.updatable,
  });
  const mappingTableInserts = Object.entries(staticArrayProperties).map(([name, definition]) =>
    defineMappingTableInsertsForArrayProperty({ name, definition, entityName: entity.name }),
  );

  // define the sql
  return `
  -- find or create the static entity
  SET v_static_id = (
    SELECT id
    FROM ${entity.name}
    WHERE 1=1
      ${uniqueStaticPropertyWhereClauseConditionals.join('\n      ')}
  );
  IF (v_static_id IS NULL) THEN -- if entity could not be found originally, create the static entity
    INSERT INTO ${entity.name}
      (uuid, ${staticPropertyColumnNames.join(', ')})
      VALUES
      (uuid(), ${staticPropertyColumnValueReferences.join(', ')});
    SET v_static_id = (
      SELECT id
      FROM ${entity.name}
      WHERE 1=1
        ${uniqueStaticPropertyWhereClauseConditionals.join('\n        ')}
    );${
      // ensure that no newlines are added if no mapping table inserts are needed
      mappingTableInserts.length ? ['', ...mappingTableInserts].join('\n\n    ') : ''
    }
  END IF;
`.trim();
};
