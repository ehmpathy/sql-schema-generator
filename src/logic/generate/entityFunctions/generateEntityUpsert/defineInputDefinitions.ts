import { prop } from '../../../../contract/module';
import { Entity } from '../../../../types';
import { castPropertyToFunctionInputDefinition } from './utils/castPropertyToFunctionInputDefinition';

export const defineInputDefinitions = ({ entity }: { entity: Entity }) => {
  // cast properties to input definitions
  const inputDefinitionsFromProperties = Object.entries(entity.properties).map((entry) =>
    castPropertyToFunctionInputDefinition({ name: entry[0], definition: entry[1] }),
  );

  // if entity is unique on uuid, then uuid must be an input
  const entityIsUniqueOnUuid = entity.unique.length === 1 && entity.unique[0] === 'uuid';
  if (entityIsUniqueOnUuid) {
    return [
      castPropertyToFunctionInputDefinition({ name: 'uuid', definition: prop.UUID() }),
      ...inputDefinitionsFromProperties,
    ];
  }

  // otherwise, then just the properties are sufficient for inputs
  return inputDefinitionsFromProperties;
};
