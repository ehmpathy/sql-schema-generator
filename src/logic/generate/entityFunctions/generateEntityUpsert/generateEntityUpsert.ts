import { Entity } from '../../../../types';
import { defineDeclarations } from './defineDeclarations';
import { defineFindOrCreateStaticEntityLogic } from './defineFindOrCreateStaticEntityLogic';
import { defineInsertVersionIfDynamicDataChangedLogic } from './defineInsertVersionIfDynamicDataChangedLogic';
import { defineUpsertCurrentVersionPointerIfNeededLogic } from './defineUpsertCurrentVersionPointerIfNeededLogic';
import { castPropertyToFunctionInputDefinition } from './utils/castPropertyToFunctionInputDefinition';

/*
1. define procedure / function (try function if possible)
2. define inputs
3. declare needed inputs
4. logic
  1. find or create the static entity, get the id of the static entity
    - if inserting, insert into mapping tables too
  2. check if the version has changed
  3. if the version has changed, insert a new version
  5. return the id of the static entity
*/

/*
  what is affected by array properties:
    - input type (json array) -vs- saved type (char(36) <- hash) && mapping table inputs
    - input name -vs- table name

  ---

  and

  ----

  so:
  - conversions:
    - 1. input name -> stored name for array properties
    - 2. normal name -> stored name for array properties
    - 3. normal name -> mapping table for array properties

  ----

  ideally:

  upsert_x(in_tag_ids VARCHAR(5000)) {
    -- calculate hashes for arrays
    v_tag_ids_hash = SHA256(in_tag_ids);
    v_name_ids_hash = SHA256(in_name_ids);

    -- find unique
    SELECT * WHERE
      ...
      AND tag_ids_hash = SHA256(v_tag_ids_hash)

    -- insert if dne
    if (null) {
      -- insert main
      INSERT INTO (tag_ids_hash)
      VALUES
        (v_tag_ids_hash)

      -- insert into mapping table, if array props exist
      FOR tag IN in_tags {
        INSERT INTO x_to_tag (x_id, tag_id) VALUES (v_entity_id, tag.id);
      }
    }

    -- do the same for the version

  }

  SO changes:
    - calculate the hash value one time for each array prop
    - change the property name if array and change the property type if array (name => ${name}_hash, type => CHAR(32))
      - when checking or inserting
    - insert into mapping tables any time inserting static row or version row
*/

/*

  // define the property names
  const updatablePropertyNames = Object.entries(entity.properties)
    .filter((entry) => !!entry[1].updatable)
    .map((entry) => entry[0]);

  // define where clause conditionals
  const updateablePropertyWhereClauseConditionals = updatablePropertyNames.map((name) =>
    propertyNameToWhereClauseConditional({ name, entity }),
  );

*/

export const generateEntityUpsert = ({ entity }: { entity: Entity }) => {
  // define the input definitions
  const inputDefinitions = Object.entries(entity.properties).map((entry) =>
    castPropertyToFunctionInputDefinition({ name: entry[0], definition: entry[1] }),
  );

  // define the declarations
  const declarations = defineDeclarations({ entity });

  // define all of the different logic required for a complete upsert
  const findOrCreateStaticEntityLogic = defineFindOrCreateStaticEntityLogic({ entity });
  const insertVersionIfDynamicDataChangedLogic = defineInsertVersionIfDynamicDataChangedLogic({ entity });
  const upsertCurrentVersionPointerIfNeededLogic = defineUpsertCurrentVersionPointerIfNeededLogic({ entity });

  // combine the version and static logic into full upsert function
  const definition = `
CREATE FUNCTION \`upsert_${entity.name}\`(
  ${inputDefinitions.join(',\n  ')}
) RETURNS bigint(20)
BEGIN
  -- declarations
  ${declarations.join('\n  ')}

  ${[findOrCreateStaticEntityLogic, insertVersionIfDynamicDataChangedLogic, upsertCurrentVersionPointerIfNeededLogic]
    .filter((sql) => !!sql)
    .join('\n\n  ')}

  -- return the static entity id
  return v_static_id;
END
  `.trim();
  return {
    name: `upsert_${entity.name}`,
    sql: definition,
  };
};
