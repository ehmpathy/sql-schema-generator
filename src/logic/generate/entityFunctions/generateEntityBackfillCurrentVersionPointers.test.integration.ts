import uuid from 'uuid/v4';
import { pg as prepare } from 'yesql';

import { DatabaseConnection, getDatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { getShowCreateFunction } from '../../../__test_utils__/getShowCreateFunction';
import { Entity } from '../../../types';
import * as prop from '../../define/defineProperty';
import { generateEntityTables } from '../entityTables/generateEntityTables';
import { generateEntityBackfillCurrentVersionPointers } from './generateEntityBackfillCurrentVersionPointers';
import { generateEntityUpsert } from './generateEntityUpsert/generateEntityUpsert';

describe('generateEntityBackfillCurrentVersionPointers', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDatabaseConnection();
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  describe('versioned entity', () => {
    const car = new Entity({
      name: 'car',
      properties: {
        vin: prop.VARCHAR(255),
        name: {
          ...prop.VARCHAR(255),
          updatable: true,
          nullable: true, // i.e., some people don't name their cars
        },
        wheels: {
          ...prop.SMALLINT(),
          updatable: true,
        },
      },
      unique: ['vin'],
    });
    beforeAll(async () => {
      // provision the table
      const tables = await generateEntityTables({ entity: car });
      await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.currentVersionPointer!.name};` });
      await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.version!.name};` });
      await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.static.name};` });
      await dbConnection.query({ sql: tables.static.sql });
      await dbConnection.query({ sql: tables.version!.sql });
      await dbConnection.query({ sql: tables.currentVersionPointer!.sql });
    });
    const upsertCar = async ({ vin, name, wheels }: { vin: string; name?: string; wheels: number }) => {
      const result = await dbConnection.query(
        prepare(`
        SELECT * FROM upsert_${car.name}(
          :vin,
          :name,
          :wheels
        );
      `)({
          vin,
          name,
          wheels,
        }),
      );
      return result.rows[0].id;
    };
    const recreateTheUpsertMethod = async () => {
      const { name, sql: upsertSql } = generateEntityUpsert({ entity: car });
      await dbConnection.query({ sql: `DROP FUNCTION IF EXISTS ${name}` });
      await dbConnection.query({ sql: upsertSql });
      return { name, sql: upsertSql };
    };
    const recreateTheBackfillMethod = async () => {
      const { name, sql: upsertSql } = generateEntityBackfillCurrentVersionPointers({ entity: car });
      await dbConnection.query({ sql: `DROP FUNCTION IF EXISTS ${name}` });
      await dbConnection.query({ sql: upsertSql });
      return { name, sql: upsertSql };
    };
    const backfillCurrentVersionPointers = async (params?: { limit: number }) => {
      const result = await dbConnection.query(
        prepare(`
          SELECT backfill_${car.name}_cvp (:limit) as rows_affected;
        `)(params || { limit: 100 }),
      );
      return result.rows[0].rows_affected as number;
    };
    const getEntityVersions = async ({ id }: { id: number }) => {
      const result = await dbConnection.query(
        prepare(`
        select * from ${car.name}_version where ${car.name}_id = :id order by created_at asc
      `)({ id }),
      );
      return result.rows;
    };
    const getEntityCurrentVersionPointer = async ({ id }: { id: number }) => {
      const result = await dbConnection.query(
        prepare(`
        select * from ${car.name}_cvp where ${car.name}_id = :id
      `)({ id }),
      );
      return result.rows;
    };
    it('should produce the same syntax as the SHOW CREATE FUNCTION query', async () => {
      const { sql, name } = await recreateTheBackfillMethod();
      const showCreateSql = await getShowCreateFunction({ dbConnection, func: name });
      expect(sql).toEqual(showCreateSql);
      expect(sql).toMatchSnapshot();
    });
    it('should find nothing needs backfill after upsert', async () => {
      await recreateTheBackfillMethod();
      await recreateTheUpsertMethod();

      // insert the car
      const props = {
        vin: uuid(),
        name: 'Bessy',
        wheels: 4,
      };
      await upsertCar(props);

      // check that rows affected by backfill = 0
      const rowsAffectedByBackfill = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfill).toEqual(0);

      // update the car
      await upsertCar({ ...props, wheels: 5 });

      // check that rows affected are still = 0
      const rowsAffectedByBackfillNow = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfillNow).toEqual(0);
    });
    it('should insert new records into cvp table, if for some reason it was never added in the first place', async () => {
      await recreateTheBackfillMethod();
      await recreateTheUpsertMethod();

      // insert the car
      const props = {
        vin: uuid(),
        name: 'Bessy',
        wheels: 4,
      };
      const id = await upsertCar(props);

      // show that before deleting, nothing was needed to be done
      const rowsAffectedByBackfill = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfill).toEqual(0);

      // delete all the cvp records
      await dbConnection.query({ sql: `delete from ${car.name}_cvp where ${car.name}_id = ${id}` });

      // prove that the target record was affected
      const rowsAffectedByBackfillNow = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfillNow).toEqual(1);

      // show that the current version pointer is correct
      const currentVersions = await getEntityVersions({ id });
      expect(currentVersions.length).toEqual(1); // we expect this in the test
      const currentVersionId = currentVersions[0].id;
      const [currentVersionPointer] = await getEntityCurrentVersionPointer({ id });
      expect(currentVersionPointer[`${car.name}_version_id`]).toEqual(currentVersionId);
    });
    it('should update records in cvp table, if for some reason the record is out of sync', async () => {
      await recreateTheBackfillMethod();
      await recreateTheUpsertMethod();

      // insert the car
      const props = {
        vin: uuid(),
        name: 'Bessy',
        wheels: 4,
      };
      const id = await upsertCar(props);

      // show that before inserting new version manually, nothing was needed to be done
      const rowsAffectedByBackfill = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfill).toEqual(0);

      // insert a new version manually
      await dbConnection.query({
        sql: `
        INSERT INTO car_version
        (car_id, name, wheels)
        VALUES
        (${id}, 'the world famous twenty-one wheeler', 21);
      `,
      });

      // prove that one pointer needs to be fixed now
      const rowsAffectedByBackfillNow = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfillNow).toEqual(1);

      // show that the current version pointer is now correct
      const currentVersions = await getEntityVersions({ id });
      expect(currentVersions.length).toEqual(2); // we expect this in the test
      const currentVersionId = currentVersions[1].id;
      const [currentVersionPointer] = await getEntityCurrentVersionPointer({ id });
      expect(currentVersionPointer[`${car.name}_version_id`]).toEqual(currentVersionId);
    });

    const upsertNewCar = async () => {
      const props = {
        vin: uuid(),
        name: 'Bessy',
        wheels: 4,
      };
      return await upsertCar(props);
    };
    const deleteCvpRecordForCarById = async ({ id }: { id: number }) => {
      await dbConnection.query({ sql: `delete from ${car.name}_cvp where ${car.name}_id = ${id}` });
    };
    it('should respect the limit on inserts', async () => {
      await recreateTheBackfillMethod();
      await recreateTheUpsertMethod();

      // upsert 5 cars
      const idsOfNewCars = [
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
      ];

      // show that before deleting, nothing was needed to be done
      const rowsAffectedByBackfill = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfill).toEqual(0);

      // manually update each row, ensuring that ids are out of sync
      await Promise.all(idsOfNewCars.map(async (id) => deleteCvpRecordForCarById({ id })));

      // show that limit is respected
      const rowsAffectedByBackfillNow = await backfillCurrentVersionPointers({ limit: 3 });
      expect(rowsAffectedByBackfillNow).toEqual(3);

      // and double prove it by showing that the remainder will be backfilled on running it again
      const rowsAffectedByBackfillNowAgain = await backfillCurrentVersionPointers({ limit: 1000 });
      expect(rowsAffectedByBackfillNowAgain).toEqual(2);
    });

    const manuallyChangeVersionOfCarById = async ({ id }: { id: number }) => {
      await dbConnection.query({
        sql: `
        INSERT INTO car_version
        (car_id, name, wheels)
        VALUES
        (${id}, 'the world famous twenty-one wheeler', 21);
      `,
      });
    };
    it('should respect the limit on updates', async () => {
      await recreateTheBackfillMethod();
      await recreateTheUpsertMethod();

      // upsert 5 cars
      const idsOfNewCars = [
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
      ];

      // show that before updating, nothing was needed to be done
      const rowsAffectedByBackfill = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfill).toEqual(0);

      // manually update each row, ensuring that ids are out of sync
      await Promise.all(idsOfNewCars.map(async (id) => manuallyChangeVersionOfCarById({ id })));

      // show that limit is respected
      const rowsAffectedByBackfillNow = await backfillCurrentVersionPointers({ limit: 3 });
      expect(rowsAffectedByBackfillNow).toEqual(3);

      // and double prove it by showing that the remainder will be backfilled on running it again
      const rowsAffectedByBackfillNowAgain = await backfillCurrentVersionPointers({ limit: 1000 });
      expect(rowsAffectedByBackfillNowAgain).toEqual(2);
    });
    it('should respect the limit on combinations of inserts and updates', async () => {
      await recreateTheBackfillMethod();
      await recreateTheUpsertMethod();

      // upsert 5 cars
      const idsOfNewCarsToDeleteRecordsFor = [
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
      ];
      const idsOfNewCarsToUpdate = [
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
        await upsertNewCar(),
      ];

      // show that before updating, nothing was needed to be done
      const rowsAffectedByBackfill = await backfillCurrentVersionPointers();
      expect(rowsAffectedByBackfill).toEqual(0);

      // manually update each row, ensuring that ids are out of sync
      await Promise.all(idsOfNewCarsToDeleteRecordsFor.map(async (id) => deleteCvpRecordForCarById({ id })));
      await Promise.all(idsOfNewCarsToUpdate.map(async (id) => manuallyChangeVersionOfCarById({ id })));

      // show that limit is respected
      const rowsAffectedByBackfillNow = await backfillCurrentVersionPointers({ limit: 7 });
      expect(rowsAffectedByBackfillNow).toEqual(7);

      // and double prove it by showing that the remainder will be backfilled on running it again
      const rowsAffectedByBackfillNowAgain = await backfillCurrentVersionPointers({ limit: 1000 });
      expect(rowsAffectedByBackfillNowAgain).toEqual(3);
    });
  });
});
