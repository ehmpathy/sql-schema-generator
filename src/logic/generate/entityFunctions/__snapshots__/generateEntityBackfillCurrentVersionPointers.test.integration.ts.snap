// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEntityBackfillCurrentVersionPointers versioned entity should produce the same syntax as the SHOW CREATE FUNCTION query 1`] = `
"CREATE FUNCTION \`backfill_car_cvp\`(in_limit INT) RETURNS int(20)
BEGIN
  -- declarations
  DECLARE v_cvp_rows_inserted INT;
  DECLARE v_cvp_rows_updated INT;
  DECLARE v_remaining_limit INT;

  -- 1. insert every time cvp dne for an entity
  INSERT INTO car_cvp
    (car_id, car_version_id)
  SELECT
    e.id,
    e_to_cv.current_version_id
  FROM car e
  JOIN (
    SELECT
      e.id as id,
      v.id as current_version_id
    FROM car e
    JOIN car_version v ON v.car_id = e.id
    WHERE 1=1
      AND effective_at = ( -- and is the currently effective version
        SELECT MAX(effective_at)
        FROM car_version ssv
        WHERE ssv.car_id = e.id -- for same entity
      )
  ) AS e_to_cv ON e_to_cv.id = e.id
  WHERE 1=1
    AND NOT EXISTS (
      SELECT 'x'
      FROM car_cvp cvp
      WHERE cvp.car_id = e.id
    )
  LIMIT in_limit;
  SET v_cvp_rows_inserted = ROW_COUNT();
  SET v_remaining_limit = in_limit - v_cvp_rows_inserted;

  -- 2. update every time cvp exists but is out of sync for an entity
  UPDATE car_cvp cvp
  JOIN (
    SELECT
      e.id as id,
      v.id as current_version_id
    FROM car e
    JOIN car_version v ON v.car_id = e.id
    WHERE 1=1
      AND effective_at = ( -- and is the currently effective version
        SELECT MAX(effective_at)
        FROM car_version ssv
        WHERE ssv.car_id = e.id -- for same entity
      )
  ) AS e_to_cv ON e_to_cv.id = cvp.car_id
  SET
    updated_at = CURRENT_TIMESTAMP(6),
    cvp.car_version_id = e_to_cv.current_version_id -- set pointer table to have current version
  WHERE cvp.car_id IN (
    SELECT id FROM ( -- ugly? agreed: https://stackoverflow.com/a/12620023/3068233 - its the best way found so far to limit in an update
      SELECT
        cvp.car_id as id
      FROM car_cvp cvp
      JOIN (
    SELECT
      e.id as id,
      v.id as current_version_id
    FROM car e
    JOIN car_version v ON v.car_id = e.id
    WHERE 1=1
      AND effective_at = ( -- and is the currently effective version
        SELECT MAX(effective_at)
        FROM car_version ssv
        WHERE ssv.car_id = e.id -- for same entity
      )
      ) AS e_to_cv ON e_to_cv.id = cvp.car_id
      WHERE 1=1
        AND cvp.car_version_id <> e_to_cv.current_version_id -- where the pointer table does not already have current version
      LIMIT v_remaining_limit
    ) as tmp
  );
  SET v_cvp_rows_updated = ROW_COUNT();

  -- return the number of rows affected
  RETURN v_cvp_rows_inserted + v_cvp_rows_updated;
END"
`;
